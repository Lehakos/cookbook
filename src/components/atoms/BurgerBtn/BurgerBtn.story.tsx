import React from 'react';
import { storiesOf } from '@storybook/react';

import BurgerBtn from './BurgerBtn';

const stories = storiesOf('Atoms|BurgerBtn', module);

stories.add('Default', () => <BurgerBtn ariaControls="main-nav" />);
