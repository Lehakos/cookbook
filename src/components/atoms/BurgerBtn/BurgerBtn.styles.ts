import styled, { css } from 'styled-components';
import { size } from 'polished';
import { ifProp } from 'styled-tools';

export const Wrapper = styled.button`
  ${props => size(props.theme.sizes[2])}

  position: relative;

  border: none;
  padding: ${props => `${props.theme.space[5]}px ${props.theme.space[4]}px`};

  background: none;
  cursor: pointer;
`;

const lineStyle = css`
  position: absolute;
  left: 0;

  width: 100%;
  height: 2px;

  background: currentColor;
`;

export const Line = styled.span`
  ${lineStyle}

  top: 50%;

  transform: translateY(-50%);
`;

export const Box = styled.span<{ isOpen: boolean }>`
  position: relative;

  display: block;
  width: 100%;
  height: 100%;

  transition-duration: 0.22s;
  transition-timing-function: cubic-bezier(0.55, 0.055, 0.675, 0.19);

  &::before,
  &::after {
    ${lineStyle}

    content: '';
  }

  &::before {
    top: 0;

    transition: top 0.1s 0.25s ease-in, opacity 0.1s ease-in;
  }

  &::after {
    bottom: 0;

    transition: bottom 0.1s 0.25s ease-in,
      transform 0.22s cubic-bezier(0.55, 0.055, 0.675, 0.19);
  }

  ${ifProp(
    'isOpen',
    css`
      transform: rotate(225deg);

      transition-delay: 0.12s;
      transition-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);

      ${Line} {
      }

      &::before {
        top: 50%;

        opacity: 0;
        transform: translateY(-50%);

        transition: top 0.1s ease-out, opacity 0.1s 0.12s ease-out;
      }

      &::after {
        bottom: 50%;

        transform: translateY(50%) rotate(-90deg);

        transition: bottom 0.1s ease-out,
          transform 0.22s 0.12s cubic-bezier(0.215, 0.61, 0.355, 1);
      }
    `,
  )}
`;
