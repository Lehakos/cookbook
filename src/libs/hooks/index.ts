export { useControlledProp } from './use-controlled-prop';
export { useElementSize } from './use-element-size';
