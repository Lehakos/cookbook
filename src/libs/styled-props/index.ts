export { typography, TypographyProps } from './typography';
export { layout, LayoutProps } from './layout';
export { flexbox, FlexboxProps } from './flexbox';
