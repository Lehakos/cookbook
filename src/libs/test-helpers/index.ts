export { render } from './render';
export { mockElementWithBoundingRect } from './mock-element-with-bounding-rect';
