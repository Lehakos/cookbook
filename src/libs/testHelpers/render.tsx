import React, { ReactElement } from 'react';
import { render as baseRender } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from 'shared/theme';

export const render = (el: ReactElement) =>
  baseRender(<ThemeProvider theme={theme}>{el}</ThemeProvider>);
