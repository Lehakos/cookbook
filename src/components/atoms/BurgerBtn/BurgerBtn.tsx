import React from 'react';

import { useControlledProp } from 'libs/useControlledProp';
import { StrictUnion } from 'shared/types';

import * as s from './BurgerBtn.styles';

type BaseProps = {
  /**
   * Описание того что делает эта кнопка для скринридеров.
   * @default 'Menu'
   */
  ariaLabel?: string;
  /**
   * ID меню, с которым связана эта кнопка. Смотри https://webplatform.github.io/docs/aria/attributes/aria-controls/
   */
  ariaControls: string;
  initialIsOpen?: boolean;
};

type ControlledIsOpen = {
  isOpen: boolean;
  onUpdateIsOpen: (isOpen: boolean) => void;
};

type Props = BaseProps & StrictUnion<ControlledIsOpen | {}>;

export default ({
  ariaLabel = 'Menu',
  ariaControls,
  initialIsOpen = false,
  isOpen,
  onUpdateIsOpen,
}: Props) => {
  const [isOpenState, setIsOpen] = useControlledProp<boolean>(
    initialIsOpen,
    isOpen,
    onUpdateIsOpen,
  );

  const handleClick = () => {
    setIsOpen(!isOpenState);
  };

  return (
    <s.Wrapper
      onClick={handleClick}
      aria-controls={ariaControls}
      aria-label={ariaLabel}
      aria-expanded={isOpen}
    >
      <s.Box isOpen={isOpenState}>
        <s.Line />
      </s.Box>
    </s.Wrapper>
  );
};
