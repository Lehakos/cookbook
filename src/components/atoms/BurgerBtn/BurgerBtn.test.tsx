import React from 'react';

import { render } from 'libs/testHelpers';

import BurgerBtn from './BurgerBtn';

describe('<BurgerBtn />', () => {
  it('правильно отображается', () => {
    const { container } = render(<BurgerBtn ariaControls="main-nav" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
