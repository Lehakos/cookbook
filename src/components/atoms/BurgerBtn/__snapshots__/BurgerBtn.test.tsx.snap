// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BurgerBtn /> правильно отображается 1`] = `
.c0 {
  height: 48px;
  width: 48px;
  position: relative;
  border: none;
  padding: 16px 12px;
  background: none;
  cursor: pointer;
}

.c2 {
  position: absolute;
  left: 0;
  width: 100%;
  height: 2px;
  background: currentColor;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1 {
  position: relative;
  display: block;
  width: 100%;
  height: 100%;
  -webkit-transition-duration: 0.22s;
  transition-duration: 0.22s;
  -webkit-transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
  transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
}

.c1::before,
.c1::after {
  position: absolute;
  left: 0;
  width: 100%;
  height: 2px;
  background: currentColor;
  content: '';
}

.c1::before {
  top: 0;
  -webkit-transition: top 0.1s 0.25s ease-in,opacity 0.1s ease-in;
  transition: top 0.1s 0.25s ease-in,opacity 0.1s ease-in;
}

.c1::after {
  bottom: 0;
  -webkit-transition: bottom 0.1s 0.25s ease-in, -webkit-transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  -webkit-transition: bottom 0.1s 0.25s ease-in, transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  transition: bottom 0.1s 0.25s ease-in, transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
}

<button
  aria-controls="main-nav"
  aria-label="Toggle menu"
  class="c0"
>
  <span
    class="c1"
  >
    <span
      class="c2"
    />
  </span>
</button>
`;
